Improve rpc-tests.py

jnewbery opened the pull request:
rpc-tests.py was converted from a shell script into python in https://github.com/bitcoin/bitcoin/pull/6616 and it shows. There are several idioms which make sense for a shell script, but are considered bad practice in a python module. This PR cleans up some of the quirks.

This pull request combines several commits which improve the code style of rpc-tests.py and move away from bad practices. Reviews/comments for any or all of the commits are welcomed:

- commit 1 replaces the tests_config module and `from tests_config import *` wildcard import with a .ini file and the Python standard library configparser module. Benefits are:
    - config lives in config files and code lives in source files. Duh.
    - avoids a wildcard import, which are usually considered bad practice because it's difficult to know where names are binded.
    - replaces some of the ugly workarounds in https://github.com/bitcoin/bitcoin/pull/8133 - doesn't leave a compiled tests_config.pyc file lying around, doesn't add qa/pull_tester to the python path, allows rpc_tests.py to be run from anywhere.
- commit 2 replaces the roll-your-own argument parsing code with Python3's standard library argparse module. Benefits are:
    - more compact, cleaner code
    - passing unknown arguments down to the individual test scripts and selecting which individual test scripts to run is now cleaner
    - it should be easier and less bug-prone to add new arguments in future
- commit 3 makes a few miscellaneous improvements to rpc-test.py's arguments:
    - make all arguments start with double dash for consistency. I don't expect those arguments were being called by automatic tools, so this shouldn't break any build/test processes.
    - improve help text and output
    - add nozmq argument to explicitly exclude the ZMQ tests
- commit 4 removes global variables and places logic in main(). Benefits:
    - allows rpc-tests.py to be imported and runtests() to be called by external modules.

[EDIT: removed commit hashes since I've now rebased]

jnewbery committed:
Use configparser in rpc-tests.py

Remove the use of wildcard imports in rpc-tests.py and replace with
configparser.

jnewbery committed:
Use argparse in rpc_tests.py

This commit replaces the roll-your-own argument parsing in rpc_tests.py
with Python's standard library argparse.

jnewbery committed:
Improve rpc-tests.py arguments

A few miscellaneous improvements to rpc-tests.py command line arguments:

 - make all arguments start with double dash for consistency
 - improve help text and output
 - add nozmq argument to explicitly exclude the ZMQ tests
 - change 'parallel' to 'jobs'

MarcoFalke commented:
Great! Concept ACK on all of those. I will take a detailed look later.

jtimon commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -184,26 +182,30 @@
-        for t in testScripts + testScriptsExt:
-            if t in opts or re.sub(".py$", "", t) in opts:
+        test_list = []
+        for t in test_scripts:
```
One shouldn't be required to pass --extended to be able to run specific tests from the extended set. You can leave the following here:
```
for t in BASE_SCRIPTS + EXTENDED_SCRIPTS:
```
Mhmm, that would still exclude zmq_test.py because of the other change you made...

jnewbery replied:
Agree. I've fixed this by adding a new ALL_SCRIPTS variable which is the concatenation of the three other lists.

isle2983 commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -30,6 +31,25 @@
+parser.add_argument('--coverage', action='store_true')
+parser.add_argument('-extended', action='store_true')
+parser.add_argument('--help', '-h', '-?', action='store_true')
+parser.add_argument('--parallel', type=int, default=4)
```
If we are considering changing this around, my vote would be for renaming `--parallel` to be `--jobs` and also giving it a '-j' alias. The advantage is that it matches `make` where most people know what `-j3` means at a glance.

jnewbery replied:
agree. Changed.

isle2983 commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -30,6 +31,25 @@
+    if arg[:2] == "--":
+        passon_args.append(arg)
+    else:
+        tests.add(arg)
```
Suggest doing this chunk with list/iterator comprehension like so:

```
tests = set(arg for arg in unknown_args if arg[:2] != "--")
passon_args = [arg for arg in unknown_args if arg[:2] == "--"]
```

jnewbery replied:
much nicer. Thanks!

isle2983 commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -184,26 +182,30 @@
+        test_list = []
+        for t in test_scripts:
+            if t in tests or re.sub(".py$", "", t) in tests:
                 test_list.append(t)
```
Suggest building test_list with list comprehension:
```
test_list = [t for t in test_scripts if
             (t in tests or re.sub(".py$", "", t) in tests)]
```

jnewbery replied:
yep. Fixed.

jtimon commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -32,12 +24,18 @@
+parser.add_argument('--help', '-h', '-?', action='store_true', help='print help text and exit')
+parser.add_argument('--nozmq', action='store_true', help='do not run the zmq tests')
+parser.add_argument('--parallel', type=int, default=4, help='how many test scripts to run in parallel. Default=4.')
+parser.add_argument('--win', action='store_true', help='signal that this is running in a Windows environment and that we should run the tests')
```
" and that we should run the tests" sounds weird to me. Are there special tests only for windows? Otherwise we just should run the tests, as always, just taking into account the windows environment.
If there's windows specific tests, what about "signal that this is running in a Windows environment and the windows specific tests sould be run"?

jnewbery replied:
I haven't changed the text in this argument (the text was previously in the docstring for the module). There are no windows specific tests - this argument simply means "I know I'm running on windows and I definitely want to run the tests. I haven't dug into why this argument was added originally, but feel free to open a new PR if you think it can be removed.

jtimon commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -23,166 +23,144 @@
-        raise
-
-BASE_SCRIPTS= [
-    # longest test should go first, to favor running tests in parallel
```
NACK on changing the order and removing the time comments (ie reverting #9276 )

jnewbery replied:
Oops. This was a bad manual merge from a old branch.

Thanks for catching this. I've fixed this so #9276 is included.

jtimon commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -198,30 +176,52 @@ def runtests():
         sys.exit(0)
 
-    coverage = None
+    runtests(test_list, config["environment"]["SRCDIR"], config["environment"]["BUILDDIR"], config["environment"]["EXEEXT"], args.parallel, args.coverage, passon_args)
```
Maybe stupid question...not sure how main() can call runtests() without having declared it yet.

jnewbery replied:
main is called in the final line of the module:

```
if __name__ == '__main__':
    main()
```

by which point all the functions have been declared. I believe this is a fairly common python module structure.

jtimon commented:
Wouldn't f1e8e33 be clearer if https://github.com/bitcoin/bitcoin/blob/master/qa/pull-tester/tests_config.py.in was removed in the same commit?

utACK modulo comments.

jnewbery commented:
@isle2983 @jtimon thanks for the reviews. I've addressed all comments in rebased/squashed commits.

@jtimon - I've removed tests_config.py.in in the first commit (which changes all the commit hashes).

isle2983 commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -179,49 +118,125 @@
-            if t in tests or re.sub(".py$", "", t) in tests:
-                test_list.append(t)
+        test_list = [t for t in test_scripts if 
+                     (t in ALL_SCRIPTS or re.sub(".py$", "", t) in ALL_SCRIPTS)]
```
This is incorrect and it prevents you from running a single test at a time e.g. `$ qa/pull-tester/rpc-tests.py segwit`

`test_scripts` can only be made up of `BASE_SCRIPTS`, `ZMQ_SCRIPTS`, and `EXTENDED_SCRIPTS`. `ALL_SCRIPTS` is all of those, so `test_list` will then always be set to all of `test_scripts`.

I believe it should be:
```
test_list = [t for t in test_scripts if (t in tests or re.sub(".py$", "", t) in tests)]
```

jnewbery replied:
You're completely right. I put the ALL_SCRIPTS in the wrong place.

Should be fixed now.

isle2983 commented:
ACK 117e0f2da241e8491b17a80b0acc34ac245d3842

Tested on Debian 8 environment with normal `./configure LDFLAGS="-L$BDB_BUILD/lib" CPPFLAGS="-I/$BDB_BUILD/include"`:

- with and without `--coverage`
- with and without `--extended`
- with and without `--nozmq`
- individual tests given as args
- `-j12` and with default jobs

Did not test:

- `passon_args`
- `--win`

jtimon commented:
Bike-shedding history, I would have preferred that ALL_SCRIPTS is declared and used in https://github.com/bitcoin/bitcoin/pull/9657/commits/0eeb09ecaee969c0f9636b893c0c3a7283006459 otherwise the commit is broken. The comprehensive lists in https://github.com/bitcoin/bitcoin/pull/9657/commits/117e0f2da241e8491b17a80b0acc34ac245d3842 could also be squashed there. It makes sense to keep s/parallel/jobs separated.
I'm not sure is worth it for you to fix the order in which things are done at this point, but I wanted to mention it.
utACK 117e0f2

jnewbery commented:
thanks @jtimon - you're right about that historic commit being broken. I did try to merge back and rebase, but it caused merge conflicts, so I didn't bother continuing.

If the maintainers really think this is important, I can go back and fix history, but these commits should be merged as a set, so I don't think it matters.

jnewbery commented:
Apologies for the review churn. I've just realised that my last commit introduced a redundant check on the names of the tests to run. I've removed that in commit https://github.com/bitcoin/bitcoin/pull/9657/commits/1bd87a63d36c79e5a019bc44e07a5a34d2bd51ea . Once review is complete I'll squash that into the previous commit.

MarcoFalke commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -187,44 +118,126 @@
+    Help text and arguments for individual test script:''',
+                                     formatter_class=argparse.RawTextHelpFormatter)
+    parser.add_argument('--coverage', action='store_true', help='generate a basic coverage report for the RPC interface')
+    parser.add_argument('--extended', action='store_true', help='run the extended test suite in addition to the basic tests')
```
The rationale in choosing `-extended` was that args with a single leading dash would only be meant for this script, and args with two dashes are meant to be passed on to the actual test scripts.

MarcoFalke commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -187,44 +118,126 @@
+    if config["environment"]["EXEEXT"] == ".exe" and not args.win:
+        # https://github.com/bitcoin/bitcoin/commit/d52802551752140cf41f0d9a225a43e84404d3e9
+        # https://github.com/bitcoin/bitcoin/pull/5677#issuecomment-136646964
+        print("Win tests currently disabled by default.  Use -win option to enable")
```
`-win` or `--win`?

MarcoFalke commented:
> so I don't think it matters.

It will break git bisect. Maybe you squash all commits that are problematic? Though, I won't block the pull based on this.

jnewbery committed:
Refactor rpc-tests.py

- add main()
- remove global variables

jnewbery commented:
> It will break git bisect. 

Good point. I've now fixed up the history so rpc-tests.py should run at all commits. Branch is now at 3de3ccdfaaa5edcf4e0924860d960e7e0db5a9b9 . The only difference from the previous branch pointer 117e0f2da241e8491b17a80b0acc34ac245d3842 (ACK'ed by @jtimon and @isle2983 ) is that I've removed a redundant if test, changed a variable name from `run_parallel` to `jobs` and added the double dash to the help text for `win`.

jtimon commented:
re-utACK 3de3ccdfaaa5edcf4e0924860d960e7e0db5a9b9
Thanks for cleaning up the history.

ryanofsky commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -184,26 +176,39 @@
-    elif len(opts) == 0 or (len(opts) == 1 and "-win" in opts):
-        test_list = testScripts
+    # Build list of tests
+    if len(tests) != 0:
```
Should avoid if len in python, https://www.python.org/dev/peps/pep-0008/

```
Yes: if not seq:
     if seq:

No: if len(seq):
    if not len(seq):
```

More occurrences a few places below, too.

jnewbery replied:
fixed

ryanofsky commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -184,26 +176,39 @@
+        # Individual tests have been specified. Run specified tests that exist
+        # in the ALL_SCRIPTS list. Accept the name with or without .py extension.
+        test_list = [t for t in ALL_SCRIPTS if
+                (t in tests or re.sub(".py$", "", t) in tests)]
```
Seems like "t in tests or " is redundant.

jnewbery replied:
This is required if the user specifies a test with the .py suffix, eg:

`rpc-tests.py rpcnamedargs.py`

ryanofsky commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -29,26 +30,22 @@
+config = configparser.ConfigParser()
+config.read_file(open(os.path.dirname(__file__) + "/tests_config.ini"))
+
+ENABLE_WALLET = config["components"]["ENABLE_WALLET"] == "True"
```
Maybe just use the config object directly below instead of defining these variables, since they are only used once. This will cut down on the number of global variables in this script, and make it clearer from the context where the settings are used where the settings come from.

jnewbery replied:
yep - I set this as a global variable to match the existing layout. It's been moved to be a local variable in the subsequent refactor commit.

I could remove the variable entirely and access the config directly in the if test, but I don't think it makes much difference at this point.

ryanofsky commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -32,12 +24,18 @@
+parser.add_argument('--help', '-h', '-?', action='store_true', help='print help text and exit')
+parser.add_argument('--jobs', '-j', type=int, default=4, help='how many test scripts to run in parallel. Default=4.')
+parser.add_argument('--nozmq', action='store_true', help='do not run the zmq tests')
+parser.add_argument('--win', action='store_true', help='signal that this is running in a Windows environment and that we should run the tests')
```
Since you're adding a dash to this argument maybe consider renaming it something comprehensible like --force. And changing help text to "Run tests even on platforms where they are disabled by default (e.g. windows)."

The "signal that this is running in a windows environment" just comes off as strange, since the script obviously knows what environment it is running in.

jnewbery replied:
Yes. @jtimon made the same point that the help text was nonsense. Good point about changing the name of the parameter since I'm already breaking backward compatibility :)

I'll change this to --force and update the help text.

ryanofsky commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -23,69 +23,8 @@
-        raise
-
 BASE_SCRIPTS= [
+    # Scripts that are run by the travis build process
```
These two lines look like one sentence, would add punctuation.

ryanofsky commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -201,30 +197,47 @@ def runtests():
-    coverage = None
+    runtests(test_list, config["environment"]["SRCDIR"], config["environment"]["BUILDDIR"], config["environment"]["EXEEXT"], args.jobs, args.coverage, passon_args)
+
+def runtests(test_list, src_dir, build_dir, exeext, jobs=1, enable_coverage=False, args=[]):
```
Maybe call it run_tests for pep8.

ryanofsky reviewed (commented):
Nice changes. Left some comments but they are minor. The one change I would really like to see is removing the DEFAULT section from config.ini, though.

ryanofsky commented:
utACK 3de3ccdfaaa5edcf4e0924860d960e7e0db5a9b9

jnewbery commented on qa/pull-tester/tests_config.ini.in:
```diff
@@ -0,0 +1,25 @@
+# These environment variables are set by the build process and read by
+# rpc-tests.py
+
+[DEFAULT]
```
Great! I wasn't aware of the get() method. Agree that removing the defaults section is much better.

jnewbery commented on qa/pull-tester/tests_config.ini.in:
```diff
@@ -0,0 +1,25 @@
+
+[components]
+# Which components are enabled. These are commented out by `configure` if they were disabled when running config.
+@ENABLE_WALLET_TRUE@ENABLE_WALLET=True
```
Good point. I'll change these to `true` and use the `getboolean()` method

jnewbery commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -30,6 +31,19 @@
+parser.add_argument('--help', '-h', '-?', action='store_true')
+parser.add_argument('--parallel', type=int, default=4)
+parser.add_argument('-win', action='store_true')
+(args, unknown_args) = parser.parse_known_args()
```
removed

jnewbery committed:
Various review markups for rpc-tests.py improvements

jnewbery commented:
Thanks for the thorough review @ryanofsky . I've applied most of those markups in a new commit. I haven't squashed them into the separate commits since there was quite a lot of code churn within those commits and it'd be tedious to get the history right.

maflcko reviewed (commented):
utACK a6a3e58c483700bd5f2bc7d117a621e0c4af60e3

There might be some minor issues (see feedback), but I think this is ready for merge nonetheless, as it comes with some great features that make future changes to the test framework a lot easier.

The feedback may be addressed in a separate pull, otherwise it would slow down progress in other test related pulls.

MarcoFalke commented:
Thanks @ryanofsky for suggesting the `getboolean()` method, this is indeed cleaner code.

jnewbery commented:
Thanks @MarcoFalke . I'll make sure to track that feedback and address it early next week.
